Cookbook (#36)
Open-EO/openeo.org: documentation/1.0/cookbook/README.md

# OpenEO Cookbook

This is the openEO cookbook that you can refer to to get a first idea on how to solve problems with the openEO API in the three client languages Python, R and JavaScript. It describes how to implement simple use cases in a pragmatic way.

Please refer to the getting started guides for [JavaScript](https://openeo.org/documentation/1.0/javascript), [Python](https://openeo.org/documentation/1.0/python/) and [R](https://openeo.org/documentation/1.0/r/) if you have never worked with the openEO API. This guide requires you to have a basic idea of how to establish a connection to a backend and how to explore that backend.

::: tip References
* [openEO processes documentation](https://openeo.org/documentation/1.0/processes.html)
* [openEO hub](https://openeo.org/documentation/1.0/processes.html) to discover backends with available data and processes
* [openEO web editor](https://editor.openeo.org) to visually build process graphs and execute jobs
:::

## Connecting to a Backend

Although already covered by the getting started guides, we want to go through our examples from the very beginning. So let's see how to connect to a backend of choice.

> <CodeSwitcher>

Author: @jonathom It's not required to list the languages any longer. They are now defined centrally in the VuePress config. You only need to list them if you derive from the pre-defined 3 languages.

# OpenEO Cookbook

> This is the openEO cookbook that you can refer to to get a first idea on how to solve problems with the openEO API in the three client languages Python, R and JavaScript. It describes how to implement simple use cases in a pragmatic way.

Author: ```suggestion
This is the openEO cookbook that you can refer to to get a first idea on how to solve problems with openEO in the three client languages Python, R and JavaScript. It describes how to implement simple use cases in a pragmatic way.
```

> Please refer to the getting started guides for [JavaScript](https://openeo.org/documentation/1.0/javascript), [Python](https://openeo.org/documentation/1.0/python/) and [R](https://openeo.org/documentation/1.0/r/) if you have never worked with the openEO API. This guide requires you to have a basic idea of how to establish a connection to a backend and how to explore that backend.

Author: ```suggestion
Please refer to the getting started guides for [JavaScript](https://openeo.org/documentation/1.0/javascript), [Python](https://openeo.org/documentation/1.0/python/) and [R](https://openeo.org/documentation/1.0/r/) if you have never worked with one of the openEO client libraries before. This guide requires you to have a basic idea of how to establish a connection to a backend and how to explore that backend.
```

::: tip References

> * [openEO processes documentation](https://openeo.org/documentation/1.0/processes.html)

Author: ```suggestion
* [openEO processes documentation](../processes.md)
```

> * [openEO hub](https://openeo.org/documentation/1.0/processes.html) to discover backends with available data and processes

Author: ```suggestion
* [openEO Hub](https://hub.openeo.org) to discover backends with available data and processes
```

> * [openEO web editor](https://editor.openeo.org) to visually build process graphs and execute jobs

Author: ```suggestion
* [openEO Web Editor](https://editor.openeo.org) to visually build process graphs and execute jobs
```

:::

Throughout these guides, code examples for all three client languages are given. Select your preferred language with the code switcher on the right-hand side to set all examples to that language.

## Connecting to a Backend

> Click the link below to see how to connect to backend via openID. You can call the connection object `con` to avoid confusion throughout the rest of the tutorials.

Author: ```suggestion
Click the link below to see how to connect to a back-end via openID Connect. You can call the connection object `con` to avoid confusion throughout the rest of the tutorials.
```

<CodeSwitcher>
<template v-slot:py>

> [Getting started: openID authentication](https://openeo.org/documentation/1.0/python/#openid-connect-authentication)

Author: ```suggestion
[Getting started: Authentication](https://openeo.org/documentation/1.0/python/#openid-connect-authentication)
```

</template>

<template v-slot:r>

> [Getting started: openID authentication](https://openeo.org/documentation/1.0/r/#openid-connect-authentication)

Author: ```suggestion
[Getting started: Authentication](https://openeo.org/documentation/1.0/r/#openid-connect-authentication)
```

</template>

<template v-slot:js>

> [Getting started: openID authentication](https://openeo.org/documentation/1.0/javascript/#openid-connect-authentication)

Author: ```suggestion
[Getting started: Authentication](https://openeo.org/documentation/1.0/javascript/#openid-connect-authentication)
```

</template>

</CodeSwitcher>

In R and JavaScript it is very useful to assign a graph-building helper object to a variable, to easily access all openEO processes and add them to the process graph that you will be building. These objects will be used throughout this guide.

<CodeSwitcher>
<template v-slot:py>

No graph-building object is required. In python, all processes are just appended to each other.

</template>
<template v-slot:r>

```r
# get a process graph builder, see ?processes
p <- processes()
```

**Note:** In all R code, `p` is used to select openEO processes.

</template>
<template v-slot:js>

```js

> // get process builder help

Author: What is this line useful for?

Reviewer: I don't understand what you mean here, please specify.

Author: "// get process builder help" - I don't understand this comment. What help are you getting here?

Reviewer: it refers to the text above:
> useful to assign a graph-building helper object to a variable, to easily access all openEO processes and add them to the process graph that you will be building

an object from which to choose openEO processes.

Author: I still think this is confusing. Is this a typo and should be  `// get process builder helper`? Anyway, the comment is not very useful IMHO.
Either remove it or make it more explicit, e.g. `// assign the graph-building helper object to "builder" for easy access to all openEO processes`

Reviewer: thanks, I just used your suggestion :+1:

var builder = await con.buildProcess();
```

**Note:** In all JavaScript code, `builder` is used to select openEO processes.

</template>
</CodeSwitcher>

## Input: `load_collection`

Before loading a collection, we need to find out the exact name of a collection we want to use (backend-specific, see references [at the top](#openeo-cookbook)). We assign the spatial and temporal extent to variables, so that we can re-use them on other collections we might want to load. Let's look for a Sentinel 2 (preprocessed level 2A preferably) collection and load the first near-infrared band (band 8). Check with the backend (refer to collection description) for the correct naming of bands (`B08` vs `B8`).

<CodeSwitcher>
<template v-slot:py>

```python
# make dictionary, containing bounding box

> brussel = {"west":4.2369, "south":50.7816, "east":4.5277, "north":50.9305}

Author: It's called "brussels" ;-)

# make list, containing the temporal interval
t = ["2020-06-01", "2020-09-01"]

# load first datacube
cube_s2_b8 = con.load_collection(
    "SENTINEL2_L2A_SENTINELHUB",
    spatial_extent = brussel,
    temporal_extent = t,
    bands = ["B08"]
)
```

</template>
<template v-slot:r>

```r
# create variables for loading collection
brussel <- list(west=4.2369, south=50.7816, east=4.5277, north=50.9305)
t <- c("2020-06-01", "2020-09-01")

# load first datacube
cube_s2_b8 <- p$load_collection(
  id = "SENTINEL2_L2A_SENTINELHUB",
  spatial_extent = iceland,
  temporal_extent = t,
  bands=c("B08")
)
```

</template>
<template v-slot:js>

```js
// make spatial and temporal extent
let brussel = {"west":4.2369, "south":50.7816, "east":4.5277, "north":50.9305};
let t = ["2020-06-01", "2020-09-01"];   

// load first cube
var cube_s2_b8 = builder.load_collection(
    "SENTINEL2_L2A_SENTINELHUB",
    brussel,
    t,
    ["B08"]
);
```

</template>
</CodeSwitcher>

## Temporal Mean: `reduce_dimension`

> We are interested in the mean of band 8 over the loaded time period, meaning we want to calculate the mean value of all timesteps per pixel and dispose of the `time` dimension afterwards (since it is empty anyway). This can be done via `reduce_dimension()`. The function requires a reducer, in our case a `mean` process, and the dimension over which to reduce, given as a string (`"t"`).

Author: It would be useful here to link to a set of reducers. We should think about implementing something like that in Process DocGen and clients.

Reviewer: I agree, just don't know how or what exactly that is. can talk about it.

Author: Yes, that was more a note for me than a to-do for you.

<CodeSwitcher>
<template v-slot:py>

```python
# reduce all timesteps
# mean_time() is a shortcut function
cube_s2_b8_red = cube_s2_b8.mean_time()

# alternatively, 'reduce_dimension' can be used
cube_s2_b8_red = cube_s2_b8.reduce_dimension(dimension="t", reducer="mean")
```

**Note:** In python, the predefined openEO function can be given as a string.

</template>
<template v-slot:r>

```r
# reduce dimension, use p$mean to give reducer
cube_s2_b8_red <- p$reduce_dimension(data = cube_s2_b8, reducer = p$mean, dimension = "t")
```

**Note:** `p$mean` means that we are using the predefined `mean` function provided by openEO.

</template>
<template v-slot:js>

```js
// reduce dimension, use builder$mean to give reducer

> var cube_s2_b8_red = builder.reduce_dimension(data = cube_s2_b8, reducer = builder.mean, dimension = "t");

Author: Wow, does that actually work? I'd expect you need to define a callback...

Reviewer: Yeah it works, same in R.
Callback might be necessary when it comes to parameters, haven't tested that yet.
Should I always stick to the callback method?

Author: No, if that works, great! Just found it funny that I implemented the  library and couldn't remember that it works so easily :-D Seems to be good work lol

Reviewer: turns out earthengine doesn't like this at all and doesn't even accept the /POST job, so this seems to be a VITO thing. I'll add proper callbacks.

Author: Interesting. Where have you found the corresponding code?

Reviewer: You mean the /POST http 500 response? Or what exactly? the 500 error is printed right into console upon execution of the script.

Author: No, I mean where have you found that `reducer = builder.mean` should work? I guess you found that in some tutorial?

Reviewer: Ah sorry, no I just tried that by myself.

```

**Note:** `builder.mean` means that we are using the predefined `mean` function provided by openEO.

</template>
</CodeSwitcher>

## Output: `save_result`

> To get a result, we first need to create a `save_result` node, in which we state the desired output format and potential parameters, both dependent on the backend you are connected to. The output formats and their parameters can e.g. be explored via the web editor: top-right corner, button "i Server".

Author: The latest version of the web editor shows them directly on the left side:

![image](https://user-images.githubusercontent.com/8262166/115871124-6a1bc880-a440-11eb-99f8-1198b47e23ec.png)

We then proceed to send that job to the backend, _without executing it_. Refer to the getting started guides on how to process results as batch or synchronous jobs. The way it is stated here allows us to log in to the web editor and look at, change, and execute the job from there.

### Raster Formats: GTiff, NetCDF, PNG

In the example, GeoTiff files are produced. You can also replace `GTiff` with `NetCDF`.

<CodeSwitcher>
<template v-slot:py>

```python
# save using save_result, give format as string
res = cube_s2_b8_red.save_result(format = "GTiff")
# send job to backend, do not execute
job = res.send_job(title = "mean_cube_py")
```

</template>
<template v-slot:r>

```r
# use list_file_formats() to be able to choose from a list
formats <- list_file_formats()

# save using save_result, give format via list
res <- p$save_result(data = cube_s2_b8_red, format = formats$output$GTiff)

# send job to backend

> job <- create_job(graph = res, title = "mean_cube_r")

Author: Maybe use a more human-readable title.

```

</template>
<template v-slot:js>

```js
// save using save_result, give fomat as string
result = builder.save_result(data = cube_s2_b8_red, format = "GTiff");

// send job to backend, but don't execute yet
var job = await con.createJob(result, "mean_cube_js");
```

</template>
</CodeSwitcher>

For a `PNG` output, we need to create the three channels red, green and blue.

### Text Formats: *JSON, CSV

> ## Troubleshooting

Author: I'd recommend putting empty chapters and todos not into the readme, so we can merge at any time.

# send job to backend
job <- create_job(graph = res, title = "mean_cube_r")
```

</template>
<template v-slot:js>

```js
// save using save_result, give fomat as string
result = builder.save_result(data = cube_s2_b8_red, format = "GTiff");

// send job to backend, but don't execute yet
var job = await con.createJob(result, "mean_cube_js");
```

</template>
</CodeSwitcher>

For a `PNG` output, we need to create the three channels red, green and blue.

### Text Formats: *JSON, CSV

## Troubleshooting

Only a collection for now.

- is the result properly saved?
- what is actually in the result? does it contain too many timesteps or bands? has the result been aggregated?
- have bands been renamed or are you trying to merge three cubes that all have the same band name? (of course monthly aggregation usuall done diffrently (agg_temp))
- parameters that are named dependent on backend: collection name, exact band name, output format (?), 

You have **feedback or** noticed an **error**? Feel free to open an issue in the [github repository](https://github.com/Open-EO/openeo.org) or use the [other communication channels](https://openeo.org/contact.html)

> * add filter bands to show python .bands and other filtr_bands?

Author: That looks like a left-over.

Reviewer: I'm thinking about this: When I wanna show PNG output in the same document here, I can either show a time-RGB (haven't got `aggregate-temporal(-frequency)` to work yet, might also be too much so early) or a bands-RGB. If doing bands, I could either load a new collection or load a bigger one in the beginning and then also explain `filter_bands` already. Doing the latter is my fav right now cause its hopefully less confusing.
Then also, PNG output a bit of a mystery at the moment, many backend differences and I don't know what they do in the background.

Author: The latter sounds more useful right now, indeed. PNG indeed seems to be pretty different across providers, but that shouldn't be an issue for the cookbook?

Reviewer: OK. I'll give it a go. Don't wanna recommend something that's then absent or different depending on backend. But I'll see about that.

Reviewer: PNG done, text output to go.

All clients have more or less different specifics when defining a callback. As you can observe directly below, one way to define a callback is to pass the callback function directly inside the "parent"-function.

For a more clean way to define a callback, see the chapter below.
:::

<CodeSwitcher>
<template v-slot:py>

```python
# reduce all timesteps
# mean_time() is a shortcut function
cube_s2_b8_red = cube_s2_b8.mean_time()

# alternatively, 'reduce_dimension' can be used
cube_s2_b8_red = cube_s2_b8.reduce_dimension(dimension="t", reducer="mean")

# additionally, reduce second collection
cube_s2_b234_red = cube_s2_b234.mean_time()
```

**Note:** In python, the callback function can be a string.

</template>
<template v-slot:r>

```r
# reduce dimension, first collection
cube_s2_b8_red <- p$reduce_dimension(data = cube_s2_b8, reducer = function(data, context) { p$mean(data) }, dimension = "t")

# reduce, second collection
cube_s2_b234_red <- p$reduce_dimension(data = cube_s2_rgb, reducer = function(data, context) { p$mean(data) }, dimension = "t")
```

**Note:** In R, we can select a callback from the `p` helper object.

</template>
<template v-slot:js>

```js
// reduce dimension

> var cube_s2_b8_red = builder.reduce_dimension(data = cube_s2_b8, reducer = new Formula("mean(data)"), dimension = "t");

Author: @me Check whether `data => this.mean(data)` works

Author: Just implemented:

![image](https://user-images.githubusercontent.com/8262166/116588054-5e864100-a91b-11eb-80c8-146aaba014ad.png)

Author: So this can be something like:
```suggestion
var cube_s2_b8_red = builder.reduce_dimension(data = cube_s2_b8, reducer = (data, _, child) => child.mean(data), dimension = "t");
```

Reviewer: changed it where applicable, added describtions :heavy_check_mark:
